### ISSUE TITLE
Fennec migration: Open tabs are duplicated

### ISSUE THREAD

**Issue by ebalazs-sv** [mozilla-mobile/android-components#5042]
## Preconditions:
Have Firefox Release installed, logged in and 2-3 opened tabs.
## Steps to reproduce
1. Replace Fennec with Fenix using a build from https://github.com/mozilla-mobile/fenix/wiki/Fennec-Migration;
2. Launch Fenix;
3. Tap on Start browsing;
4. Observe the open tabs.
### Expected behavior
The opened tabs are displayed.
### Actual behavior
The opened tabs are duplicated. The visited pages are also duplicated in history.
### Notes:
The number of duplicated tabs is related with the numbers of time the app crashes (bug https://github.com/mozilla-mobile/android-components/issues/5041).
![Screenshot_20191113-103153](https://user-images.githubusercontent.com/48799384/68750856-4593ae00-0609-11ea-9056-473d30929e64.png)
### Device information

* Android device: Google Pixel 2 (Android 9) - ARM64, ARM.
* Fenix version: Nightly 191109 GV 71

**Comment by st3fan** [comment]
Is it also a bug that we import `about:home` or is that fine?

**Comment by ebalazs-sv** [comment]
From documentation "Default bookmarks will be migrated but may not work (about:firefox, about:add-ons)". 
@pocmo Is this also the case for about:home? Is it better to file a bug for the about page being migrated but not usable? 
Thanks!

**Comment by pocmo** [comment]
Right now that is the expected state, we migrate open tabs as-is. There's no incentive to make the about pages work. We could consider filtering them out when migrating tabs. But I think that has a low priority since that is really not the end of the world.

**Comment by ebalazs-sv** [comment]
@pocmo Hi! This is not reproducible on today's Release migration builds 12/19 with  Google Pixel 2 (Android 9), Samsung Galaxy S8 (Android 9), Huawei Honor 8 (Android 7.0), Xiaomi Mi4i (Android 5.0.2). 

I have noticed that this is not reproducible since https://github.com/mozilla-mobile/android-components/issues/5041 and the related issues were fixed. 

This is still reproducible on Beta and Nightly migration builds, see https://github.com/mozilla-mobile/android-components/issues/5276.

**Comment by pocmo** [comment]
I recently noticed something similar, also after a crash that happened while migrating. We do protect against some crashes, but not others. This can lead us crashing before we can safe the migration state to disk. Which will cause another migration to run. And in the case of "open tabs" we'd migrate again and duplicate them.

There are two things we can do:
* Maybe not migrate tabs if there are already tabs. This is what we do for logins for example. However this will make it harder to re-run a migration - but we can work around that.
* Make sure we protect against all crashes during migration and save (partial) state as early as possible.

**Comment by grigoryk** [comment]
At least we'll persist states of migrations right after they run, not at the very end.

**Comment by ebalazs-sv** [comment]
I had tested this on Release, Beta, Nightly migration builds 01/13 with Samsung Galaxy S8 (Android 9), Huawei Honor 8 (Android 7.0), Xiaomi Mi4i (Android 5.0.2) and the open tabs are not duplicated.
I will move this to Done. Thanks!
